lib/components/T3Form/T3Form.vue

```diff
@@ -0,0 +1,145 @@
+    />
+    <form
+      ref="form"
+      class="t3-form"
+      novalidate
+      @submit.prevent="onSubmit"
+      @reset="resetForm"
+    >
+      <T3FormFieldList
+        ref="fieldlist"
+        :elements="elements"
+        :components="components"
+        :classes="classes"
+        :i18n="i18n"
+        @change="updateModel"
+      />
+      <T3FormCta
+        :state="state"
+        :i18n="i18n"
+      />
+      <slot
+        name="after"
+        :validator="validator"
+        :state="state"
+      />
+    </form>
+  </ValidationObserver>
+</template>
+<script>
+import { ValidationObserver } from 'vee-validate'
+import extendableComponents from '../../mixins/component/extendableComponents'
+import T3FormFieldList from './T3FormFieldList/T3FormFieldList.vue'
+import T3FormCta from './T3FormCta/T3FormCta.vue'
+
+const defaultState = {
+  sending: null,
+  error: null,
+  failure: null,
+  success: null
+}
+
+export default {
+  name: 'T3Form',
+  components: {
+    ValidationObserver,
+    T3FormFieldList,
+    T3FormCta
+  },
+  mixins: [extendableComponents],
+  props: {
+    elements: {
+      type: Array,
+      required: true
+    },
+    components: {
+      type: [Object, Boolean],
+      default: false
+    },
+    classes: {
+      type: [Object, Boolean],
+      default: false
+    },
+    labels: {
+      type: Object,
+      default: () => ({})
+    }
+  },
+  data () {
+    return {
+      model: {},
+      state: { ...defaultState },
+      i18n: {
+        submitButton: 'Submit',
+        sendingLabel: 'Sending',
+        resetButton: 'Reset form',
+        serverSuccess: 'The form was sent, thank you.',
+        serverError: 'We can not process form right now, please try again later.',
+        validationErrors: 'There were some errors, review the form'
+      }
+    }
+  },
+  computed: {
+    validatorRef () {
+      return this.$refs.validator
+    },
+    fieldListRef () {
+      return this.$refs.fieldlist
+    }
+  },
+  created () {
+    this.i18n = Object.freeze(Object.assign(this.i18n, this.labels))
+  },
+  methods: {
+    updateModel (model) {
+      this.model = Object.assign(this.model, model)
+    },
+    focusOnErrors () {
+      const errors = Object.entries(this.validatorRef.refs)
+        .map(([key, value]) => ({
+          key,
+          value
+        }))
+        .filter((error) => {
+          return (!error || !error.value || !error.value.failedRules) ? false : Object.keys(error.value.failedRules).length > 0
+        })
+      if (errors && errors.length > 0) {
+        this.validatorRef.refs[errors[0].key].$el.scrollIntoView({
+          behavior: 'smooth',
+          block: 'center'
+        })
+      }
+    },
+    setState (state) {
+      this.state = { ...defaultState, [state]: true }
+    },
+    resetForm () {
+      this.fieldListRef.restoreModel()
+      this.validatorRef.reset()
+    },
+    onSubmit () {
+      this.validatorRef.validate().then((success) => {
```

Comment:
The Async function is more readable in my opinion. I always recommend to use it instead of `.then`.

Reply:
I'm callback man xD but you are right, we should use async/await :D